kernel/Kconfig

```diff
@@ -833,15 +833,29 @@ config KERNEL_MEM_POOL
 
 if KERNEL_MEM_POOL
 
+config HEAP_MEM_POOL_MIN_SIZE
+	int
+	default 1024 if POSIX_MQUEUE
+	default 0
```

Comment:
This is a step in the right direction, but it's a fake solution, i.e. it gives the impression that it can be set to the minimum heap required for functionality but will be trumped by other Kconfigs.

Assume an application enables 3 features, one sets this to 3KB, another set it to 0, another sets it to 1MB - what is the actual value you get here? It's actually undefined, because it depends on the order in which the Kconfigs are encountered. Therefore with this huge issue, I cannot +1 this idea

Reply:
> This is a step in the right direction, but it's a fake solution, i.e. it gives the impression that it can be set to the minimum heap required for functionality but will be trumped by other Kconfigs.
> 
> Assume an application enables 3 features, one sets this to 3KB, another set it to 0, another sets it to 1MB - what is the actual value you get here? It's actually undefined, because it depends on the order in which the Kconfigs are encountered. Therefore with this huge issue, I cannot +1 this idea

Yes, that's correct. I was aware of this more fundamental issue when I filed #65722 and after the discussion there proceeded to create this PR. AFAIK there is no existing mechanism in the Zephyr build system currently to select the maximum from a set of multiple build-time defined values (whether it's Kconfig or something else). If there is, then that would definitely be a better solution to the problem.

The situation as it stands today isn't quite as bad as you describe though, since the defaults are not being set "all over the place" by "features", rather it's currently only boards or SoCs that set default values. So in that sense I do think this PR is taking the situation in a better direction. E.g. right now there are 11 tests applications under `tests/kernel` that all set explicit heap sizes that are too small for any ACPI-using board. The apps build fine, but result in runtime failures with misleading error messages on all of these boards.

Reply:
It's introducing a known defect to try and address an issue, so instead of the original 1 issue, you now have 2 issues. Agreed, this cannot be fixed in Kconfig as it stands today, I do not have a proposal for how this can be addressed, just that this is not the way to solve it

Reply:
Yep, basically what this does is to group things to "the application" and "everything else (kernel, subsystems, etc)". The second category is left on its own to figure out a consistent minimum value. That problem is not solved by this PR, although it kind of does work due the limited set of places where defaults are attempted to be set.

I see a few more potential complexities:
 - Should it really be `MAX(A, B, C, D)`? What if the submodules run together and at the same time need successful memory allocations? Wouldn't it then be `A + B + C + D` that we need?
 - It's not strictly that component A always has a specific requirement, rather the requirement might depend on what other components a component is combined with. E.g. ACPI always requires some substantial amount of heap, but the exact amount depends on what platform you use it on (it can range from a few hundred kilobytes to several megabytes).

Reply:
@nordicjm aggregating a list of different minimum system heap size requirements at build time, and then picking the highest value (also at build time), could be done using a cmake global property, i.e. each place (SoC, board, subsystem, etc) that has a specific minimum requirement would add something like the following to their CMakeLists.txt file:
```cmake
set_property(GLOBAL APPEND PROPERTY HEAP_MEM_POOL_REQ 32768)
```

This would replace the `HEAP_MEM_POOL_MIN_SIZE` Kconfig option I've so far been proposing. Does that sound like an acceptable solution for you?

Reply:
> ```cmake
> set_property(GLOBAL APPEND PROPERTY HEAP_MEM_POOL_REQ 32768)
> ```
> 

The above could also be made slightly nicer looking by using a helper function, assuming it's ok to add such a thing to `cmake/modules/extensions.cmake`. In that case all places needing to declare their requirement would just do something like:
```cmake
zephyr_heap_mem_pool_req(32768)
```

Reply:
Drive-by comment: I like the idea, but as someone who fantasizes about a cmake-free build future, I don't really love the idea of adding another dependency.  Also cmake is surprisingly opaque about "how to tell if this dependency is actually going to be built", IIRC.  It's not enough to do it by subsystem, you need to know if it was actually linked.

Where we've wanted to do this in the past (collecting settings/metadata from unrelated subsystems based on whether they're built or not) it's worked pretty well to rely on the linker.  Consider a special section containing heap requirements  that we could parse out of an intermediate link step the same way we do for SYS_INIT or IRQ_CONNECT declarations?

Reply:
Is there a way to make a Kconfig "additive"? For example, if feature X is enabled add <n> to a value? 

I've run across this type of thing with stack sizes where adding features should add to the stack size...

Reply:
> Is there a way to make a Kconfig "additive"?

No, Kconfig is so basic that you can't even have a Kconfig option that adds a number to another number (or option)

Reply:
@andyross thanks for the proposal! I'll explore that next.

Do people here have any thoughts on whether the end result should be max(a,b,c,d) or a+b+c+d? Should it perhaps be guided by a Kconfig option?

Reply:
another possibility would be to introduce a `max(a,b)` function as Kconfig extension command, like we have with `dt` functions.

That would not be a Kconfig symbol as we know it, but instead using variables, and therefore potentially more errorprone.

But posting it so we can discuss all options.
Here https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/kconfig/kconfigfunctions.py define a `max(a,b)`: 
```diff
diff --git a/scripts/kconfig/kconfigfunctions.py b/scripts/kconfig/kconfigfunctions.py
index 621ec3cb3a8..50bfcb5e520 100644
--- a/scripts/kconfig/kconfigfunctions.py
+++ b/scripts/kconfig/kconfigfunctions.py
@@ -756,6 +756,8 @@ def shields_list_contains(kconf, _, shield):
 
     return "y" if shield in list.split(";") else "n"
 
+def kconfigmax(kconf, _, a, b):
+    return max(a,b)
 
 # Keys in this dict are the function names as they appear
 # in Kconfig files. The values are tuples in this form:
@@ -807,4 +809,5 @@ functions = {
         "dt_nodelabel_array_prop_has_val": (dt_nodelabel_array_prop_has_val, 3, 3),
         "dt_gpio_hogs_enabled": (dt_gpio_hogs_enabled, 0, 0),
         "shields_list_contains": (shields_list_contains, 1, 1),
+        "max": (kconfigmax, 2, 2),
 }
```

allows us to do following in Kconfig files:
```
# Kconfig file A
HEAP_MEM_POOL_MIN_SIZE := $(max,$(HEAP_MEM_POOL_MIN_SIZE),2048)
```
and later
```
# Kconfig file B
HEAP_MEM_POOL_MIN_SIZE := $(max,$(HEAP_MEM_POOL_MIN_SIZE),8192)
```
or
```
# Kconfig file B
HEAP_MEM_POOL_MIN_SIZE := $(max,$(HEAP_MEM_POOL_MIN_SIZE),512)
```

and then in [kernel/Kconfig](https://github.com/zephyrproject-rtos/zephyr/pull/65908/files#diff-05292158cc01da0f208bda8b8a77b9bd2f2a4fdf6f259259d99f18b8688575fc) have:
```
config HEAP_MEM_POOL_SIZE
	int "Heap memory pool size (in bytes)"
	default $(HEAP_MEM_POOL_MIN_SIZE)
```

Now, the use of `HEAP_MEM_POOL_MIN_SIZE` as a variable requires that all `HEAP_MEM_POOL_MIN_SIZE := ...` assignment happens **before** `kernel/Kconfig` is sourced, but that is still far better than today where the first `default <value>` always win.

Combining this with an `add(a,b)` will even allow subsys to decide whether to do a max or add, as they seem fit.
So subsys A can do:
```
# A requires min 2048, but can live side-by-side with other modules.
HEAP_MEM_POOL_MIN_SIZE := $(max,$(HEAP_MEM_POOL_MIN_SIZE),2048)
```
but B can do:
```
# B requires additional 1024, regardless of current min.
HEAP_MEM_POOL_MIN_SIZE := $(add,$(HEAP_MEM_POOL_MIN_SIZE),1024)
```

Reply:
downsides is of course such approach is more error prone.
First, it's harder to debug as you cannot see who decided the value in case someone did this by mistake:
```
HEAP_MEM_POOL_MIN_SIZE := 128
```
The alternative of
```
config HEAP_MEM_POOL_MIN_SIZE_FOO
        int
        default 128
```
allows developers to easily see where HEAP_MEM_POOL_MIN_SIZE_FOO was defined.

Also mistakes like this can be hard to trace down:
```
HEAP_MEM_POOL_SIZE_MIN := $(max,$(HEAP_MEM_POOL_MIN_SIZE),4096)
```

Reply:
I lean mostly towards the combined Kconfig+CMake approach (https://github.com/zephyrproject-rtos/zephyr/pull/65908#discussion_r1418931869), but there is also something nice about staying purely in Kconfig.
Overall I would say 60/40 in favor of combined Kconfig+CMake approach, but would like to hear what others think.

Reply:
@tejlmand thanks for the suggestion! I think I'll go with the "Kconfig prefix" approach then.

To cater for subsystems wanting their requirement to be additive vs subsystems that can deal with sharing the heap with others, we could have two different prefixes, e.g. HEAP_MEM_POOL_MIN_SIZE_* and HEAP_MEM_POOL_ADD_SIZE_*. This would then result in a calculation max(a,b,c) + sum(x,y,z) where the first term comes from MIN_SIZE_* values and the second from ADD_SIZE_* values? Is this getting too complicated? 😄 I could in a first iteration only cover the max(a,b,c) case.